src/device_handle/async_api.rs

```diff
@@ -0,0 +1,218 @@
+            };
+            // Step 2 of async api
+            ffi::libusb_fill_bulk_transfer(
+                ptr.as_ptr(),
+                device.as_raw(),
+                endpoint,
+                result.buffer.as_ptr() as *mut u8,
+                result.buffer.len().try_into().unwrap(),
+                Self::transfer_cb,
+                closure_as_ptr.cast(),
+                timeout,
+            )
+        };
+        result
+    }
+
+    /// Submits a transfer to libusb's event engine.
+    /// # Panics
+    /// A transfer should not be double-submitted! Only re-submit after a submission has
+    /// returned Err, or the callback has gotten an Err.
+    // Step 3 of async API
+    #[allow(unused)]
+    pub fn submit(self: &mut Pin<Box<Self>>) -> Result<(), TransferError> {
+        let errno = unsafe { ffi::libusb_submit_transfer(self.ptr.as_ptr()) };
+        use ffi::constants::*;
+        match errno {
+            0 => Ok(()),
+            LIBUSB_ERROR_BUSY => {
+                panic!("Do not double-submit a transfer!")
+            }
+            LIBUSB_ERROR_NOT_SUPPORTED => Err(TransferError::Other("Unsupported transfer!")),
+            LIBUSB_ERROR_INVALID_PARAM => Err(TransferError::Other("Transfer size too large!")),
+            LIBUSB_ERROR_NO_DEVICE => Err(TransferError::Disconnected),
+            _ => Err(TransferError::Errno("Unable to submit transfer. ", errno)),
+        }
+    }
+
+    // We need to invoke our closure using a c-style function, so we store the closure
+    // inside the custom user data field of the transfer struct, and then call the
+    // user provided closure from there.
+    // Step 4 of async API
+    extern "system" fn transfer_cb(transfer: *mut ffi::libusb_transfer) {
+        // Safety: libusb should never make this null, so this is fine
+        let transfer = unsafe { &mut *transfer };
+
+        // sanity
+        debug_assert_eq!(
+            transfer.transfer_type,
+            ffi::constants::LIBUSB_TRANSFER_TYPE_BULK
+        );
+
+        // sanity
+        debug_assert_eq!(
+            std::mem::size_of::<*mut F>(),
+            std::mem::size_of::<*mut c_void>(),
+        );
+        // Safety: The pointer shouldn't be a fat pointer, and should be valid, so
+        // this should be sound
+        let closure = unsafe {
+            let closure: *mut F = std::mem::transmute(transfer.user_data);
+            &mut *closure
+        };
+
+        use ffi::constants::*;
+        match transfer.status {
+            LIBUSB_TRANSFER_CANCELLED => {
+                // Step 5 of async API: Transfer was cancelled, free the transfer
+                unsafe { ffi::libusb_free_transfer(transfer) }
+            }
+            LIBUSB_TRANSFER_COMPLETED => {
+                debug_assert!(transfer.length >= transfer.actual_length); // sanity
+                let data = unsafe {
+                    std::slice::from_raw_parts(transfer.buffer, transfer.actual_length as usize)
```

Comment:
One other safety consideration is that the transfer is no longer considered busy by libusb at the time of the callback, and the AsyncTransfer might be owned by another thread that could resubmit it during the execution of this callback or the user closure. That would illegally alias and mutate the `transfer` and buffer references here.